org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/ui/text/correction/TypeMismatchBaseSubProcessor.java

```diff
@@ -324,6 +341,62 @@ public void collectChangeSenderTypeProposals(IInvocationContext context, Express
 						proposals.add(p2);
 				}
 			}
+
+			if (nodeToCast.getNodeType() == ASTNode.CLASS_INSTANCE_CREATION) {
+				HashSet<CompletionProposal> completionProposals = new HashSet<>();
+				CompletionRequestor requestor = new CompletionRequestor(false) {
+					@Override
+					public void accept(CompletionProposal proposal) {
+						completionProposals.add(proposal);
+					}
+				};
+
+				requestor.setAllowsRequiredProposals(CompletionProposal.CONSTRUCTOR_INVOCATION, CompletionProposal.TYPE_REF, true);
+				ASTNode constructorNode = context.getCoveringNode();
+				if(!(constructorNode instanceof ClassInstanceCreation)) {
+					constructorNode = ASTNodes.getParent(constructorNode, ClassInstanceCreation.class);
+				}
+				cu.codeComplete(constructorNode.getStartPosition() + SKIP_NEW_KEYWORD, requestor, new NullProgressMonitor());
```

Comment:
I would use CompletionTimeoutProgressMonitor instead of the NullProgressMonitor. This at least puts some upper limit on how long this can take.